File: code/48-Rotate Image/48-Rotate Image.c
/*
LeetCode 48 - Rotate Image

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
*/
/*
Rotate Image
You are given an n x n 2D matrix representing an image.

Rotate the image by 90 degrees (clockwise).

Follow up:
Could you do this in-place?

就是先按照副对角线对称交换，然后按照水平中线对称交换

因为恰好是90°的关系所以计算比较方便，可以展开成逆时针90°，那就先按照主对角线对称交换
*/

void swap(int* a, int* b) {
  int tmp = *a;
  *a = *b;
  *b = tmp;
}

void rotate(int** matrix, int matrixSize, int* matrixColSize){
    for (int i = 0; i < matrixSize; i++) {
        for (int j = 0; j < matrixSize - i; j ++) {
            swap(&(matrix[i][j]), &(matrix[matrixSize - j - 1][matrixSize - i - 1]));
            // 二维数组直接计算地址方法
            // swap(*(matrix + i) + j, *(matrix + matrixSize - j - 1) + matrixSize - i - 1);
        }
    }
    for (int i = 0; i < matrixSize / 2; i ++) {
        for (int j = 0; j < matrixSize; j ++) {
            swap(&(matrix[i][j]), &(matrix[matrixSize - i - 1][j]));
            // swap(*(matrix + i) + j, *(matrix + matrixSize - i - 1) + j);
        }
    }
}

File: code/48-Rotate Image/48-Rotate Image.py
# -*- coding: utf-8 -*
"""
LeetCode 48 - Rotate Image

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
"""
r"""
@author: Kaka Chen
@license: (C) Copyright Personal Uncommercial Limited.
@contact: kakachen1990@gmail.com
@file: 48-Rotate Image.py
@time: 2020/12/19 1:30 下午
@desc:
You are given an n x n 2D matrix representing an image, rotate the image by 90 degrees (clockwise).

You have to rotate the image in-place, which means you have to modify the input 2D matrix directly. DO NOT allocate another 2D matrix and do the rotation.



Example 1:


Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
Output: [[7,4,1],[8,5,2],[9,6,3]]
Example 2:


Input: matrix = [[5,1,9,11],[2,4,8,10],[13,3,6,7],[15,14,12,16]]
Output: [[15,13,2,5],[14,3,4,1],[12,6,8,9],[16,7,10,11]]
Example 3:

Input: matrix = [[1]]
Output: [[1]]
Example 4:

Input: matrix = [[1,2],[3,4]]
Output: [[3,1],[4,2]]


Constraints:

matrix.length == n
matrix[i].length == n
1 <= n <= 20
-1000 <= matrix[i][j] <= 1000

"""


class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        for i in range(n):
            for j in range(n - i):
                matrix[i][j], matrix[n - j - 1][n - i - 1] = matrix[n - j - 1][n - i - 1], matrix[i][j]

        for i in range(n / 2):
            for j in range(n):
                matrix[i][j], matrix[n - i - 1][j] = matrix[n - i - 1][j], matrix[i][j]
